timberio/vector
# enhancement(vrl): Split VRL functions into Wasm compatible and incompatible

This PR splits the VRL standard lib into Wasm-compatible and -incompatible functions. Although this isn't *yet* useful, it will be in the future when we begin running VRL in environments outside of Vector itself.

Two VRL functions, `get_hostname` and `parse_grok`, have dependencies that make them incompatible with the `wasm32-unknown-unknown` target. The `uuid_v4` function has a dependency (the `uuid` library) with an optional `wasm-bindgen` feature that makes it Wasm friendly. The approach here relies heavily on Rust's feature system to make the `default` feature of `vrl-stdlib` behave exactly as it currently does while also providing a `wasm` feature. I can't way I *love* this approach from a cognitive clarity perspective, but it definitely "works" just fine. I'm amenable to suggestions.

To verify that this division works as expected:

```bash
cd lib/vrl/stdlib
cargo check --target wasm32-unknown-unknown --no-default-features --features wasm_compatible # SUCCESS
cargo check --target wasm32-unknown-unknown #FAILURE
```

An alternative possibility here would be adding versions of the two incompatible functions that remain technically part of the stdlib but error when called, like so:

```bash
$ get_hostname()
the get_hostname function isn't available in this environment

# Or alternatively something fun:
"your_browser"
```

That could work something like this:

```toml
# Cargo.toml
wasm_compatible = ["default"]
```

```rust
// get_hostname.rs
#[cfg(not(feature = "wasm_compatible"))]
fn resolve(&self, _: &mut Context) -> Resolved {
    Ok(hostname::get()
        .map_err(|error| format!("failed to get hostname: {}", error))?
        .to_string_lossy()
        .into())
}

#[cfg(feature = "wasm_compatible")]
fn resolve(&self, _: &mut Context) -> Resolved {
    Err("the get_hostname function isn't available in this environment")
}
```

I have no strong feelings about the choice between excluding incompatible functions and the include-but-error approach. I do lean, however, toward simply excluding incompatible functions for the sake of Least Surprise.
